README.md
[![hex.pm version](https://img.shields.io/hexpm/v/webdavex.svg?style=flat)](https://hex.pm/packages/webdavex)
[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/webdavex/)
[![Build Status](https://travis-ci.org/mugimaru73/webdavex.svg?branch=master)](https://travis-ci.org/mugimaru73/webdavex)

# Webdavex

[hackney](https://github.com/benoitc/hackney) based WebDAV client.

Webdavex aimed to work with [nginx implementation of WebDAV](https://nginx.org/en/docs/http/ngx_http_dav_module.html)
which means it does not support `PROPFIND`, `PROPPATCH`, `LOCK` and `UNLOCK` methods.

## Installation

```elixir
def deps do
  [
    {:webdavex, "~> 0.3.3"}
  ]
end
```

## Quick start

```elixir
defmodule MyApp.WebdavClient do
  use Webdavex, base_url: "https://webdav.host:888"
end

MyApp.WebdavClient.put("image.png", {:file, Path.absname("files/image.png")})
```

Refer to [Webdavex.Client](https://hexdocs.pm/webdavex/Webdavex.Client.html) API docs for more details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

## [0.3.3] - 2022-01-23

- Fix typespecs.

## [0.3.2] - 2021-12-08

- Fix compiler warnings.

## [0.3.1] - 2021-12-08

- Fix `Webdavex.Client.put/3` connections leaking.

## [0.3.0] - 2018-09-24

- Update `:hackney_options` whitelist, see `Webdavex.Config`

## [0.2.0] - 2018-09-09

### Added

- `Webdavex.Client.head/2` method
- `Webdavex.Client.get_stream/2` method

### Changed

- improve tests with dynamic `Bypass` port

## [0.1.0] - 2018-09-09

### Added

- Initial implementation.


test/webdavex_test.exs
defmodule WebdavexTest do
  use ExUnit.Case, async: false
  doctest Webdavex.Config

  alias Plug.Conn

  @image_path Path.join(["test", "support", "fixtures", "img.png"]) |> Path.absname()
  @image_content File.read!(@image_path)

  defmodule Klient do
    use Webdavex.Agent, base_url: "http://localhost"
  end

  setup do
    bypass = Bypass.open()
    {:ok, _pid} = Klient.start_link(base_url: "http://localhost:#{bypass.port}/dav", headers: [{"foo", "bar"}])
    {:ok, bypass: bypass}
  end

  defmodule RequestsTracer do
    @moduledoc "Traces sequential requests order"
    use Agent

    @spec start_link(list(String.t())) :: {:ok, pid}
    def start_link(requests) do
      Agent.start_link(fn -> requests end)
    end

    @spec requests(pid) :: list(String.t())
    def requests(pid) do
      Agent.get(pid, fn r -> r end)
    end

    @spec notify_requested(pid, String.t()) :: list(String.t())
    def notify_requested(pid, path) do
      Agent.get_and_update(pid, fn [expected_path | rest_paths] = prev_state ->
        assert path == expected_path
        {prev_state, rest_paths}
      end)
    end
  end

  defp assert_adds_default_header(conn) do
    %{headers: [{k, v}]} = Klient.config()
    assert_header(conn.req_headers, k, v)
  end

  defp assert_header(headers, key, value) do
    header = Enum.find(headers, fn {k, _v} -> k == key end)
    assert {key, value} == header
  end

  test "config/0 returns configuration struct", %{bypass: bypass} do
    assert Klient.config() == %Webdavex.Config{
             base_url: URI.parse("http://localhost:#{bypass.port}/dav"),
             hackney_options: [],
             headers: [{"foo", "bar"}]
           }
  end

  describe "put/1" do
    test "sends PUT request for a new file", %{bypass: bypass} do
      Bypass.expect_once(bypass, "PUT", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)

        {:ok, body, conn} = Conn.read_body(conn)
        assert @image_content == body

        Conn.resp(conn, 201, "")
      end)

      assert {:ok, :created} == Klient.put("images/img.png", {:file, @image_path})
    end

    test "does not leak connections", %{bypass: bypass} do
      Bypass.expect(bypass, "PUT", "/dav/images/img.png", fn conn ->
        Conn.resp(conn, Enum.random([201, 204, 500, 502, 404]), "")
      end)

      for _ <- 1..30 do
        Klient.put("images/img.png", {:file, @image_path})
      end

      assert :hackney_pool.get_stats(:default)[:in_use_count] == 0
    end

    test "returns an error if file does not exist" do
      assert {:error, :enoent} == Klient.put("images/img.png", {:file, Path.absname("foobar.png")})
    end

    test "sends PUT request for existing file", %{bypass: bypass} do
      Bypass.expect_once(bypass, "PUT", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)

        {:ok, body, conn} = Conn.read_body(conn)
        assert @image_content == body

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :updated} == Klient.put("images/img.png", {:file, @image_path})
    end

    test "sends PUT request for binary content", %{bypass: bypass} do
      Bypass.expect_once(bypass, "PUT", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)

        {:ok, body, conn} = Conn.read_body(conn)
        assert @image_content == body

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :updated} == Klient.put("images/img.png", {:binary, @image_content})
    end

    test "sends PUT request from stream", %{bypass: bypass} do
      stream = ["foo", "bar", "baz"] |> Enum.map(fn v -> v <> "\n" end)
      data = Enum.join(stream)

      Bypass.expect_once(bypass, "PUT", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)

        {:ok, body, conn} = Conn.read_body(conn)
        assert data == body

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :updated} == Klient.put("images/img.png", {:stream, stream})
    end
  end

  describe "head/1" do
    test "returns response headers", %{bypass: bypass} do
      Bypass.expect_once(bypass, "HEAD", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)

        conn
        |> Conn.put_resp_header("Foo-Bar", "baz")
        |> Conn.resp(200, "")
      end)

      assert {:ok, headers} = Klient.head("images/img.png")
      assert {"Foo-Bar", "baz"} == Enum.find(headers, fn {k, _v} -> k == "Foo-Bar" end)
    end
  end

  describe "get/1" do
    test "loads file content", %{bypass: bypass} do
      Bypass.expect_once(bypass, "GET", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)
        Conn.send_file(conn, 200, @image_path)
      end)

      assert {:ok, content} = Klient.get("images/img.png")
      assert content == @image_content
    end
  end

  describe "get_stream/1" do
    test "returns file resource", %{bypass: bypass} do
      Bypass.expect_once(bypass, "GET", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)
        Conn.send_file(conn, 200, @image_path)
      end)

      tempfile = Path.join("test", :crypto.strong_rand_bytes(12) |> Base.url_encode64()) |> Path.absname()
      on_exit(fn -> File.rm(tempfile) end)

      assert {:ok, stream} = Klient.get_stream("images/img.png")

      stream
      |> Stream.into(File.stream!(tempfile, [:write]))
      |> Stream.run()

      assert @image_content == File.read!(tempfile)
    end
  end

  describe "copy/2" do
    test "sends COPY request with default overwrite option value (true)", %{bypass: bypass} do
      Bypass.expect_once(bypass, fn conn ->
        assert conn.method == "COPY"
        assert conn.request_path == "/dav/images/img.png"
        assert_header(conn.req_headers, "overwrite", "T")
        assert_header(conn.req_headers, "destination", "#{Klient.config().base_url}/images/img_copy.png")
        assert_adds_default_header(conn)

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :copied} == Klient.copy("images/img.png", "images/img_copy.png")
    end

    test "sends COPY request with overwrite=false", %{bypass: bypass} do
      Bypass.expect_once(bypass, fn conn ->
        assert conn.method == "COPY"
        assert conn.request_path == "/dav/images/img.png"
        assert_header(conn.req_headers, "overwrite", "F")
        assert_header(conn.req_headers, "destination", "#{Klient.config().base_url}/images/img_copy.png")
        assert_adds_default_header(conn)

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :copied} == Klient.copy("images/img.png", "images/img_copy.png", false)
    end
  end

  describe "move/2" do
    test "sends MOVE request with default overwrite option value (true)", %{bypass: bypass} do
      Bypass.expect_once(bypass, fn conn ->
        assert conn.method == "MOVE"
        assert conn.request_path == "/dav/images/img.png"
        assert_header(conn.req_headers, "overwrite", "T")
        assert_header(conn.req_headers, "destination", "http://localhost:#{bypass.port}/dav/images/img_copy.png")
        assert_adds_default_header(conn)

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :moved} == Klient.move("images/img.png", "images/img_copy.png")
    end

    test "sends MOVE request with overwrite=false", %{bypass: bypass} do
      Bypass.expect_once(bypass, fn conn ->
        assert conn.method == "MOVE"
        assert conn.request_path == "/dav/images/img.png"
        assert_header(conn.req_headers, "overwrite", "F")
        assert_header(conn.req_headers, "destination", "http://localhost:#{bypass.port}/dav/images/img_copy.png")
        assert_adds_default_header(conn)

        Conn.resp(conn, 204, "")
      end)

      assert {:ok, :moved} == Klient.move("images/img.png", "images/img_copy.png", false)
    end
  end

  describe "delete/1" do
    test "sends DELETE request", %{bypass: bypass} do
      Bypass.expect_once(bypass, "DELETE", "/dav/images/img.png", fn conn ->
        assert_adds_default_header(conn)
        Conn.send_file(conn, 204, @image_path)
      end)

      assert {:ok, :deleted} = Klient.delete("images/img.png")
    end
  end

  describe "mkcol_recursive/1" do
    test "sends multiple MKCOL requests for file", %{bypass: bypass} do
      {:ok, tracer_pid} = RequestsTracer.start_link(["/dav/images", "/dav/images/foo"])

      Bypass.expect(bypass, fn conn ->
        assert conn.method == "MKCOL"
        RequestsTracer.notify_requested(tracer_pid, conn.request_path)

        assert_adds_default_header(conn)
        Conn.resp(conn, 201, "")
      end)

      assert {:ok, "/images/foo"} == Klient.mkcol_recursive("images/foo/bar.png")
      assert RequestsTracer.requests(tracer_pid) == []
    end

    test "halts after first failure and returns its result", %{bypass: bypass} do
      Bypass.expect(bypass, fn conn ->
        assert conn.method == "MKCOL"

        case conn.request_path do
          "/dav/images" ->
            Conn.resp(conn, 201, "")

          "/dav/images/foo" ->
            Conn.resp(conn, 500, "")

          path ->
            raise(RuntimeError, "It is expected that #{path} wont be requested.")
        end
      end)

      assert {:error, :http_500} == Klient.mkcol_recursive("images/foo/bar/baz/qux/")
    end

    test "sends multiple MKCOL requests for folder", %{bypass: bypass} do
      {:ok, tracer_pid} = RequestsTracer.start_link(["/dav/images", "/dav/images/foo", "/dav/images/foo/bar"])

      Bypass.expect(bypass, fn conn ->
        assert conn.method == "MKCOL"
        RequestsTracer.notify_requested(tracer_pid, conn.request_path)

        assert_adds_default_header(conn)
        Conn.resp(conn, 201, "")
      end)

      assert {:ok, "/images/foo/bar"} == Klient.mkcol_recursive("images/foo/bar/")
      assert RequestsTracer.requests(tracer_pid) == []
    end
  end

  describe "mkcol/1" do
    test "sends MKCOL request", %{bypass: bypass} do
      Bypass.expect_once(bypass, fn conn ->
        assert conn.method == "MKCOL"
        assert conn.request_path == "/dav/images"

        assert_adds_default_header(conn)
        Conn.resp(conn, 201, "")
      end)

      assert {:ok, :created} == Klient.mkcol("images")
    end
  end
end


mix.exs
defmodule Webdavex.MixProject do
  use Mix.Project

  def project do
    [
      app: :webdavex,
      version: "0.3.3",
      elixir: "~> 1.6",
      start_permanent: Mix.env() == :prod,
      description: "hackney based WebDAV client.",
      package: package(),
      deps: deps(),
      source_url: "https://github.com/mugimaru73/webdavex",
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:hackney, "~> 1.0"},
      {:ex_doc, "~> 0.14", only: :dev},
      {:credo, "~> 0.9.2", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.0.0-rc.3", only: [:dev], runtime: false},
      {:bypass, "~> 0.8", only: :test}
    ]
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README.md", "LICENSE", ".credo.exs"],
      maintainers: ["mugimaru73"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/mugimaru73/webdavex"
      }
    ]
  end

  defp docs do
    [
      main: "readme",
      extras: ["README.md", "CHANGELOG.md"]
    ]
  end
end


.credo.exs
# This file contains the configuration for Credo and you are probably reading
# this after creating it with `mix credo.gen.config`.
#
# If you find anything wrong or unclear in this file, please report an
# issue on GitHub: https://github.com/rrrene/credo/issues
#
%{
  #
  # You can have as many configs as you like in the `configs:` field.
  configs: [
    %{
      #
      # Run any config using `mix credo -C <name>`. If no config name is given
      # "default" is used.
      name: "default",
      #
      # these are the files included in the analysis
      files: %{
        #
        # you can give explicit globs or simply directories
        # in the latter case `**/*.{ex,exs}` will be used
        included: ["lib/", "test/"],
        excluded: [~r"/_build/", ~r"/deps/"]
      },
      #
      # If you create your own checks, you must specify the source files for
      # them here, so they can be loaded by Credo before running the analysis.
      requires: [],
      #
      # Credo automatically checks for updates, like e.g. Hex does.
      # You can disable this behaviour below:
      check_for_updates: true,
      #
      # If you want to enforce a style guide and need a more traditional linting
      # experience, you can change `strict` to true below:
      strict: true,
      #
      # If you want to use uncolored output by default, you can change `color`
      # to `false` below:
      #
      color: true,
      #
      # You can customize the parameters of any check by adding a second element
      # to the tuple.
      #
      # To disable a check put `false` as second element:
      #
      #     {Credo.Check.Design.DuplicatedCode, false}
      #
      checks: [
        {Credo.Check.Consistency.ExceptionNames},
        {Credo.Check.Consistency.LineEndings},
        {Credo.Check.Consistency.SpaceAroundOperators},
        {Credo.Check.Consistency.SpaceInParentheses},
        {Credo.Check.Consistency.TabsOrSpaces},

        # For some checks, like AliasUsage, you can only customize the priority
        # Priority values are: `low, normal, high, higher`
        {Credo.Check.Design.AliasUsage, priority: :low, if_called_more_often_than: 2},

        # For others you can set parameters

        # If you don't want the `setup` and `test` macro calls in ExUnit tests
        # or the `schema` macro in Ecto schemas to trigger DuplicatedCode, just
        # set the `excluded_macros` parameter to `[:schema, :setup, :test]`.
        {Credo.Check.Design.DuplicatedCode, excluded_macros: []},

        # You can also customize the exit_status of each check.
        # If you don't want TODO comments to cause `mix credo` to fail, just
        # set this value to 0 (zero).
        {Credo.Check.Design.TagTODO, exit_status: 0},
        {Credo.Check.Design.TagFIXME},

        {Credo.Check.Readability.FunctionNames},
        {Credo.Check.Readability.LargeNumbers, false},
        {Credo.Check.Readability.MaxLineLength, priority: :low, max_length: 140, ignore_specs: true},
        {Credo.Check.Readability.ModuleAttributeNames},
        {Credo.Check.Readability.ModuleDoc},
        {Credo.Check.Readability.ModuleNames},
        {Credo.Check.Readability.ParenthesesInCondition},
        {Credo.Check.Readability.PredicateFunctionNames},
        {Credo.Check.Readability.TrailingBlankLine},
        {Credo.Check.Readability.TrailingWhiteSpace},
        {Credo.Check.Readability.VariableNames},
        {Credo.Check.Readability.Specs},
        {Credo.Check.Readability.PreferImplicitTry, false},

        {Credo.Check.Refactor.ABCSize, max_size: 50},
        {Credo.Check.Refactor.CondStatements},
        {Credo.Check.Refactor.FunctionArity},
        {Credo.Check.Refactor.MatchInCondition},
        {Credo.Check.Refactor.PipeChainStart, false},
        {Credo.Check.Refactor.CyclomaticComplexity},
        {Credo.Check.Refactor.NegatedConditionsInUnless},
        {Credo.Check.Refactor.NegatedConditionsWithElse},
        {Credo.Check.Refactor.Nesting},
        {Credo.Check.Refactor.UnlessWithElse},

        {Credo.Check.Warning.IExPry},
        {Credo.Check.Warning.IoInspect},
        {Credo.Check.Warning.OperationOnSameValues},
        {Credo.Check.Warning.BoolOperationOnSameValues},
        {Credo.Check.Warning.UnusedEnumOperation},
        {Credo.Check.Warning.UnusedKeywordOperation},
        {Credo.Check.Warning.UnusedListOperation},
        {Credo.Check.Warning.UnusedStringOperation},
        {Credo.Check.Warning.UnusedTupleOperation},
        {Credo.Check.Warning.OperationWithConstantResult},

        # Custom checks can be created using `mix credo.gen.check`.
        #
      ]
    }
  ]
}


lib/webdavex.ex
defmodule Webdavex do
  @moduledoc """
  WebDAV client for elixir applications. See `Webdavex.Client`.

  Refer to `Webdavex.Config` for detailed information on available options.

  ## Examples

  ### General usage

      defmodule MyClient
        use Webdavex, base_url: "https://webdav.com/my_app"
      end

      MyClient.get("images/foobar.png")
      {:ok, ...}

  ### Dynamic config
  In order to change configuration during runtime you can wrap a client in `Agent`
  and override `config/0` to fetch data from its state.

      defmodule MyClient do
        use Webdavex.Agent, base_url: "https://webdav.com"
      end

      {:ok, _pid} = MyClient.start_link()
      MyClient.update_config(%{base_url: "http://host", headers: [{"Foo", "bar}]})

      MyClient.get("images/foobar.png")
      {:ok, ...}
  """

  alias Webdavex.Client

  defmacro __using__(opts) do
    quote do
      Module.put_attribute(__MODULE__, :config, Webdavex.Config.new(unquote(opts)))

      @spec config :: Webdavex.Config.t()
      @doc "Returns webdav client configuration."
      def config, do: @config

      @spec head(path :: String.t()) :: {:ok, list({String.t(), String.t()})} | {:error, atom}
      def head(path), do: Client.head(config(), path)

      @spec get(path :: String.t()) :: {:ok, binary} | {:error, atom}
      def get(path), do: Client.get(config(), path)

      @spec get_stream(path :: String.t()) :: {:ok, Enumerable.t()} | {:error, atom}
      def get_stream(path), do: Client.get_stream(config(), path)

      @spec put(path :: String.t(), {:file, file_path :: String.t()}) :: {:ok, :created | :updated} | {:error, atom}
      @spec put(path :: String.t(), {:binary, content :: binary}) :: {:ok, :created | :updated} | {:error, atom}
      def put(path, content), do: Client.put(config(), path, content)

      @spec move(source :: String.t(), dest :: String.t(), overwrite :: boolean) :: {:ok, :moved} | {:error, atom}
      def move(source, dest, overwrite \\ true), do: Client.move(config(), source, dest, overwrite)

      @spec copy(source :: String.t(), dest :: String.t(), overwrite :: boolean) :: {:ok, :copied} | {:error, atom}
      def copy(source, dest, overwrite \\ true), do: Client.copy(config(), source, dest, overwrite)

      @spec delete(path :: String.t()) :: {:ok, :deleted} | {:error, atom}
      def delete(path), do: Client.delete(config(), path)

      @spec mkcol(path :: String.t()) :: {:ok, :created} | {:error, atom}
      def mkcol(path), do: Client.mkcol(config(), path)

      @spec mkcol_recursive(path :: String.t()) :: {:ok, :created} | {:error, atom}
      def mkcol_recursive(path), do: Client.mkcol_recursive(config(), path)

      defoverridable config: 0
    end
  end
end


lib/webdavex/agent.ex
defmodule Webdavex.Agent do
  @moduledoc """
  Wraps `Webdavex` into `Agent` in order to change configuration during runtime.

      defmodule MyClient
        use Webdavex.Agent, base_url: "http://placeholder"
      end

      {:ok, _pid} = MyClient.start_link(base_url: "http://webdav.host")
      MyClient.update_config(base_url: "https://webdav.host", headers: [{"foo", "bar"}])
      MyClient.get("image.png")
  """

  defmacro __using__(default_config) do
    quote do
      use Agent
      use Webdavex, unquote(default_config)

      @spec start_link(Keyword.t() | map | Webdavex.Config.t()) :: {:ok, pid}
      def start_link(config \\ @config) do
        Agent.start_link(fn -> Webdavex.Config.new(config) end, name: __MODULE__)
      end

      @spec update_config(Keyword.t() | map | Webdavex.Config.t()) :: Webdavex.Config.t()
      def update_config(map) do
        Agent.update(__MODULE__, Webdavex.Config.new(map))
      end

      @spec config :: Webdavex.Config.t()
      def config do
        Agent.get(__MODULE__, fn config -> config end)
      end
    end
  end
end


lib/webdavex/config.ex
defmodule Webdavex.Config do
  @hackney_options_whitelist [:pool, :ssl_options, :connect_options, :proxy, :insecure, :connect_timeout, :recv_timeout]
  @default_headers []
  @default_hackney_options []

  @moduledoc """
  `Webdavex.Client` configuration.

  ## Options

  ### :base_url, required.
  Schema, host, port and root path of webdav endpoint.
  Example: "https://myhost.com:8081/something/webdav".

  ### :headers, default: #{inspect(@default_headers)}.
  A list of HTTTP headers that will be added to each of `Webdavex.Client` request.
  Example: [{"X-Webdav-Client", "webdavex"}].

  HTTP basic auth could be implemented using `:headers` options:
  ```
  username = "client"
  password = "supersecret"
  digest = :base64.encode(username <> ":" <> password)

  headers = [{"Authorization", "Basic " <> digest}]
  ```
  ### :hackney_options, default: #{inspect(@default_hackney_options)}.
  Options are limited to #{inspect(@hackney_options_whitelist)}, refer to `:hackney.request/5`
  [docs](https://hexdocs.pm/hackney/) for detailed information.

  ## Examples

      iex> Webdavex.Config.new(base_url: "http://myhost.com")
      %Webdavex.Config{
        base_url: URI.parse("http://myhost.com"),
        hackney_options: #{inspect(@default_hackney_options)},
        headers: #{inspect(@default_headers)}
      }

      iex> Webdavex.Config.new(
      ...>   base_url: "http://myhost.com",
      ...>   headers: [{"X-Something", "value"}],
      ...>   hackney_options: [pool: :webdav, foo: 1]
      ...> )
      %Webdavex.Config{
        base_url: URI.parse("http://myhost.com"),
        headers: [{"X-Something", "value"}],
        hackney_options: [pool: :webdav]
      }
  """

  @type t :: %__MODULE__{base_url: URI.t(), hackney_options: Keyword.t(), headers: Keyword.t()}
  defstruct [:base_url, :hackney_options, :headers]

  @spec new(map | Keyword.t() | __MODULE__.t()) :: __MODULE__.t()
  @doc "Converts enumerable into `Webdavex.Config` struct."
  def new(%__MODULE__{} = config), do: config

  def new(opts) do
    base_url = Access.get(opts, :base_url, nil) || raise(ArgumentError, "[#{__MODULE__}] `base_url` is missing.")

    struct(
      __MODULE__,
      base_url: URI.parse(base_url),
      hackney_options: Keyword.get(opts, :hackney_options, @default_hackney_options) |> filter_hackney_opts(),
      headers: Keyword.get(opts, :headers, @default_headers)
    )
  end

  defp filter_hackney_opts(opts) do
    Enum.reject(opts, fn {k, _v} -> k not in @hackney_options_whitelist end)
  end
end


lib/webdavex/client.ex
defmodule Webdavex.Client do
  @moduledoc """
  [hackney](https://github.com/benoitc/hackney) based WebDAV client.

  It is recommended to use `Webdavex` module to define your own client
  instead of using `Webdavex.Client` directly.

  With
      defmodule MyClient
        use Webdavex, base_url: "http://webdav.host"
      end

  you can write
      MyClient.get("image.png")

  instead of
      Webdavex.Config.new(base_url: "http://webdav.host") |> Webdavex.Client.get("image.png")
  """

  alias Webdavex.Config
  require Logger

  @spec head(Config.t(), path :: String.t()) :: {:ok, list({String.t(), String.t()})} | {:error, atom}
  @doc """
  Issues a HEAD request to webdav server.
  Returns response headers if target file exists.
  It migth be useful to check content type and size of file (Content-Type and Content-Length headers respectively).

  ### Examples

      MyClient.head("foo.jpg")
      {:ok, [
        {"Date", "Sun, 09 Sep 2018 18:13:00 GMT"},
        {"Content-Type", "image/jpeg"},
        {"Content-Length", "870883"},
        {"Last-Modified", "Sat, 08 Sep 2018 18:00:51 GMT"},
      ]}
  """
  def head(%Config{} = config, path) do
    case request(:head, path, [], "", config) do
      {:ok, 200, headers} ->
        {:ok, headers}

      error ->
        wrap_error(error)
    end
  end

  @spec get(Config.t(), path :: String.t()) :: {:ok, binary} | {:error, atom}
  @doc """
  Gets a file from webdav server.

  ### Examples

      MyClient.get("foo.png")
      {:ok, <<binary content>>}
  """
  def get(%Config{} = config, path) do
    with {:ok, 200, _, ref} <- request(:get, path, [], "", config),
         {:ok, body} <- :hackney.body(ref) do
      {:ok, body}
    else
      error ->
        wrap_error(error)
    end
  end

  @spec get_stream(Config.t(), path :: String.t()) :: {:ok, Enumerable.t()} | {:error, atom}
  @doc """
  Same as `get/2`, but returns content reader in form of `Stream.resource/3`.
  See `Webdavex.Helpers.Hackney.stream_body/1` and `:hackney.stream_body/1`.

  ### Examples
  Stream content into a file

      with {:ok, stream} <- MyClient.get_stream("foo.png") do
        stream
        |> Stream.into(File.stream!("/local/path/foo.png", [:write]))
        |> Stream.run
      end
  """
  def get_stream(%Config{} = config, path) do
    case request(:get, path, [], "", config) do
      {:ok, 200, _headers, ref} ->
        {:ok, Webdavex.Helpers.Hackney.stream_body(ref)}

      error ->
        wrap_error(error)
    end
  end

  @spec put(Config.t(), path :: String.t(), {:file, String.t()} | {:binary, binary} | {:stream, Enumerable.t()}) ::
          {:ok, :created} | {:ok, :updated} | {:error, atom}
  @doc """
  Uploads local file, `Stream` or binary content to webdav server.

  Note that webdav specification only allows to create files in already existing folders
  which means thah you must call `mkcol/2` or `mkcol_recursive/2` manually.

  ### Examples
  Upload local file

      MyClient.put("image.png", {:file, "/home/username/img.png"})
      {:ok, :created}

  Upload custom binary content

      MyClient.put("file.ext", {:binary, <<42, 42, 42>>})
      {:ok, :created}

  Chunked upload from a `Stream`

    MyClient.put("file.ext", {:stream, File.stream("/path/to/file.ext")})
    {:ok, :created}
  """
  def put(%Config{} = config, path, content) do
    with {:ok, ref} <- do_put(config, path, content) do
      case :hackney.start_response(ref) do
        {:ok, 204, _headers, ref} ->
          :ok = :hackney.skip_body(ref)
          {:ok, :updated}

        {:ok, 201, _headers, ref} ->
          :ok = :hackney.skip_body(ref)
          {:ok, :created}

        {:ok, _status, _headers, ref} = error ->
          :ok = :hackney.skip_body(ref)
          wrap_error(error)

        error ->
          wrap_error(error)
      end
    else
      error ->
        wrap_error(error)
    end
  end

  @spec move(Config.t(), source :: String.t(), dest :: String.t(), overwrite :: boolean) ::
          {:ok, :moved} | {:error, atom}

  @doc """
  Moves a file on webdav server.

  If overwrite option (3rd argument) is set to false (true by default) will return {:error, :http_412}
  in case of destination file already exists.

  ### Examples

      MyClient.move("imag.png", "image.png")
      {:ok, :moved}
  """
  def move(%Config{base_url: base_url} = config, source_path, destination_path, overwrite) do
    dest_header = destination_header(base_url, destination_path)
    overwrite_header = overwrite_header(overwrite)

    case request(:move, source_path, [dest_header, overwrite_header], "", config) do
      {:ok, 204, _, _ref} ->
        {:ok, :moved}

      error ->
        wrap_error(error)
    end
  end

  @spec copy(Config.t(), source :: String.t(), dest :: String.t(), overwrite :: boolean) ::
          {:ok, :copied} | {:error, atom}

  @doc """
  Copies a file on webdav server. Refer to `move/4` for details.

  ### Examples
      MyClient.copy("image.png", "image_copy.png")
      {:ok, :copied}
  """
  def copy(%Config{base_url: base_url} = config, source_path, dest_path, overwrite) do
    dest_header = destination_header(base_url, dest_path)
    overwrite_header = overwrite_header(overwrite)

    case request(:copy, source_path, [dest_header, overwrite_header], "", config) do
      {:ok, 204, _, _ref} ->
        {:ok, :copied}

      error ->
        wrap_error(error)
    end
  end

  @spec delete(Config.t(), path :: String.t()) :: {:ok, :deleted} | {:error, atom}

  @doc """
  Deletes file or directory on webdav server.

  ### Examples

      MyClient.delete("file.png")
      {:ok, :deleted}
  """
  def delete(%Config{} = config, path) do
    case request(:delete, path, [], "", config) do
      {:ok, 204, _, _} ->
        {:ok, :deleted}

      error ->
        wrap_error(error)
    end
  end

  @spec mkcol(Config.t(), path :: String.t()) :: {:ok, :created} | {:error, atom}

  @doc """
  Creates a folder on wedav server.

  Note that webdav does not allow to create nested folders (HTTP 409 will be returned).
  Use `mkcol_recursive/2` to create nested directory structure.

  ### Examples

      MyClient.mkcol("foo")
      {:ok, :created}
  """
  def mkcol(%Config{} = config, path) do
    case request(:mkcol, path, [], "", config) do
      {:ok, 201, _, _} ->
        {:ok, :created}

      error ->
        wrap_error(error)
    end
  end

  @spec mkcol_recursive(Config.t(), path :: String.t()) :: {:ok, String.t()} | {:error, atom}
  @doc """
  Creates nested folders for given path or file.

  Performs sequential `mkcol/2` calls to populate nested structure.

  ### Examples
  Create structure for a file.
      MyClient.mkcol_recursive("foo/bar/baz/file.png")
      {:ok, :created}
  "foo", "foo/bar" and "foo/bar/baz" folders will be created.

  Create structure for a folder.
      MyClient.mkcol_recursive("foo/bar/")
      {:ok, :created}
  Note that trailing slash is required for directory paths.
  """
  def mkcol_recursive(%Config{} = config, path) do
    path
    |> Path.dirname()
    |> String.split("/")
    |> Enum.reduce_while({:ok, ""}, fn folder, {:ok, prefix} ->
      new_path = prefix <> "/" <> folder

      case mkcol(config, new_path) do
        {:ok, _} ->
          {:cont, {:ok, new_path}}

        error ->
          {:halt, error}
      end
    end)
  end

  defp do_put(config, path, {:file, file_path}) do
    case File.read(file_path) do
      {:ok, content} ->
        do_put(config, path, {:binary, content})

      error ->
        error
    end
  end

  defp do_put(config, path, {mode, data}) when mode in [:stream, :binary] do
    with {:ok, ref} <- request(:put, path, [], :stream, config),
         :ok <- do_send_body(mode, ref, data),
         :ok <- :hackney.finish_send_body(ref) do
      {:ok, ref}
    else
      error ->
        wrap_error(error)
    end
  end

  defp do_send_body(:binary, ref, data), do: :hackney.send_body(ref, data)

  defp do_send_body(:stream, ref, stream) do
    Enum.reduce_while(stream, :ok, fn part, :ok ->
      case :hackney.send_body(ref, part) do
        :ok -> {:cont, :ok}
        any -> {:halt, any}
      end
    end)
  end

  defp request(meth, url, headers, body, config) do
    started_at = now()
    full_url = full_url(config.base_url, url)
    headers = config.headers ++ headers

    case :hackney.request(meth, full_url, headers, body, config.hackney_options) do
      {:ok, status, _, _} = result ->
        log_request(meth, full_url, {:ok, status}, started_at)
        result

      result ->
        log_request(meth, full_url, result, started_at)
        result
    end
  end

  defp log_request(meth, url, result, started_at) do
    Logger.debug(fn ->
      meth = meth |> to_string |> String.upcase()
      "#{meth} #{url} completed with #{inspect(result)} in #{now() - started_at}ms"
    end)
  end

  defp now, do: :erlang.system_time(:millisecond)

  defp full_url(base_url, url) when is_binary(url), do: full_url(base_url, URI.parse(url))

  defp full_url(%URI{path: base_path} = base_url, %URI{} = uri) do
    base_url
    |> URI.merge(%{uri | path: Path.join(to_string(base_path), to_string(uri.path))})
    |> to_string
  end

  defp overwrite_header(true), do: {"Overwrite", "T"}
  defp overwrite_header(false), do: {"Overwrite", "F"}

  defp destination_header(base_url, url), do: {"Destination", full_url(base_url, url)}

  defp wrap_error({:ok, status, _headers}), do: {:error, String.to_atom("http_#{status}")}
  defp wrap_error({:ok, status, _headers, _ref}), do: {:error, String.to_atom("http_#{status}")}
  defp wrap_error({:error, reason}), do: {:error, reason}
end
